In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

In [2]:
# Crear carpeta "images" si no existe
if not os.path.exists("images"):
    os.makedirs("images")

In [16]:
pd.set_option('display.max_columns', None)

## Carga y exploración inicial:

In [17]:
df_history = pd.read_csv("data/Customer Loyalty History.csv") 
#primera exploración visual
df_history

Loyalty Number Country          Province        City Postal Code  \
0              480934  Canada           Ontario     Toronto     M2Z 4K1   
1              549612  Canada           Alberta    Edmonton     T3G 6Y6   
2              429460  Canada  British Columbia   Vancouver     V6E 3D9   
3              608370  Canada           Ontario     Toronto     P1W 1K4   
4              530508  Canada            Quebec        Hull     J8Y 3Z5   
...               ...     ...               ...         ...         ...   
16732          823768  Canada  British Columbia   Vancouver     V6E 3Z3   
16733          680886  Canada      Saskatchewan      Regina     S1J 3C5   
16734          776187  Canada  British Columbia   Vancouver     V5R 1W3   
16735          906428  Canada             Yukon  Whitehorse     Y2K 6R0   
16736          652627  Canada          Manitoba    Winnipeg     R2C 0M5   

       Gender Education    Salary Marital Status Loyalty Card       CLV  \
0      Female  Bachelor   83236.0        Married         Star   3839.14   
1        Male   College       NaN       Divorced         Star   3839.61   
2        Male   College       NaN         Single         Star   3839.75   
3        Male   College       NaN         Single         Star   3839.75   
4        Male  Bachelor  103495.0        Married         Star   3842.79   
...       ...       ...       ...            ...          ...       ...   
16732  Female   College       NaN        Married         Star  61850.19   
16733  Female  Bachelor   89210.0        Married         Star  67907.27   
16734    Male   College       NaN         Single         Star  74228.52   
16735    Male  Bachelor  -57297.0        Married         Star  10018.66   
16736  Female  Bachelor   75049.0        Married         Star  83325.38   

      Enrollment Type  Enrollment Year  Enrollment Month  Cancellation Year  \
0            Standard             2016                 2                NaN   
1            Standard             2016                 3                NaN   
2            Standard             2014                 7             2018.0   
3            Standard             2013                 2                NaN   
4            Standard             2014                10                NaN   
...               ...              ...               ...                ...   
16732        Standard             2012                12                NaN   
16733        Standard             2014                 9                NaN   
16734        Standard             2014                 3                NaN   
16735  2018 Promotion             2018                 4                NaN   
16736        Standard             2015                12             2016.0   

       Cancellation Month  
0                     NaN  
1                     NaN  
2                     1.0  
3                     NaN  
4                     NaN  
...                   ...  
16732                 NaN  
16733                 NaN  
16734                 NaN  
16735                 NaN  
16736                 8.0  

[16737 rows x 16 columns]

In [22]:
df_history.tail()

Loyalty Number Country          Province        City Postal Code  \
16732          823768  Canada  British Columbia   Vancouver     V6E 3Z3   
16733          680886  Canada      Saskatchewan      Regina     S1J 3C5   
16734          776187  Canada  British Columbia   Vancouver     V5R 1W3   
16735          906428  Canada             Yukon  Whitehorse     Y2K 6R0   
16736          652627  Canada          Manitoba    Winnipeg     R2C 0M5   

       Gender Education   Salary Marital Status Loyalty Card       CLV  \
16732  Female   College      NaN        Married         Star  61850.19   
16733  Female  Bachelor  89210.0        Married         Star  67907.27   
16734    Male   College      NaN         Single         Star  74228.52   
16735    Male  Bachelor -57297.0        Married         Star  10018.66   
16736  Female  Bachelor  75049.0        Married         Star  83325.38   

      Enrollment Type  Enrollment Year  Enrollment Month  Cancellation Year  \
16732        Standard             2012                12                NaN   
16733        Standard             2014                 9                NaN   
16734        Standard             2014                 3                NaN   
16735  2018 Promotion             2018                 4                NaN   
16736        Standard             2015                12             2016.0   

       Cancellation Month  
16732                 NaN  
16733                 NaN  
16734                 NaN  
16735                 NaN  
16736                 8.0

In [24]:
df_history.sample(5)

Loyalty Number Country          Province         City Postal Code  \
9890           187902  Canada     New Brunswick  Fredericton     E3B 2H2   
3339           720131  Canada  British Columbia     Whistler     V6T 1Y8   
5946           204237  Canada  British Columbia    Vancouver     V6E 3D9   
13788          106964  Canada     New Brunswick      Moncton     E1A 2A7   
9068           272848  Canada      Saskatchewan       Regina     S6J 3G0   

       Gender Education   Salary Marital Status Loyalty Card       CLV  \
9890     Male   College      NaN         Single         Star   2141.98   
3339   Female  Bachelor  96550.0        Married       Aurora  12280.77   
5946     Male   College      NaN         Single         Nova   4595.04   
13788  Female   College      NaN         Single         Star   5568.95   
9068     Male  Bachelor  71922.0        Married         Nova  13167.61   

      Enrollment Type  Enrollment Year  Enrollment Month  Cancellation Year  \
9890         Standard             2017                 4                NaN   
3339         Standard             2018                11                NaN   
5946         Standard             2016                 1             2018.0   
13788        Standard             2018                 8                NaN   
9068         Standard             2018                 7                NaN   

       Cancellation Month  
9890                  NaN  
3339                  NaN  
5946                  3.0  
13788                 NaN  
9068                  NaN

In [18]:
df_history.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16737 entries, 0 to 16736
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Loyalty Number      16737 non-null  int64  
 1   Country             16737 non-null  object 
 2   Province            16737 non-null  object 
 3   City                16737 non-null  object 
 4   Postal Code         16737 non-null  object 
 5   Gender              16737 non-null  object 
 6   Education           16737 non-null  object 
 7   Salary              12499 non-null  float64
 8   Marital Status      16737 non-null  object 
 9   Loyalty Card        16737 non-null  object 
 10  CLV                 16737 non-null  float64
 11  Enrollment Type     16737 non-null  object 
 12  Enrollment Year     16737 non-null  int64  
 13  Enrollment Month    16737 non-null  int64  
 14  Cancellation Year   2067 non-null   float64
 15  Cancellation Month  2067 non-null   float64
dtypes: f

In [19]:
df_activity = pd.read_csv("data/Customer Flight Activity.csv")
df_activity

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
0               100018  2017      1               3                        0   
1               100102  2017      1              10                        4   
2               100140  2017      1               6                        0   
3               100214  2017      1               0                        0   
4               100272  2017      1               0                        0   
...                ...   ...    ...             ...                      ...   
405619          999902  2018     12               0                        0   
405620          999911  2018     12               0                        0   
405621          999940  2018     12               3                        0   
405622          999982  2018     12               0                        0   
405623          999986  2018     12               0                        0   

        Total Flights  Distance  Points Accumulated  Points Redeemed  \
0                   3      1521               152.0                0   
1                  14      2030               203.0                0   
2                   6      1200               120.0                0   
3                   0         0                 0.0                0   
4                   0         0                 0.0                0   
...               ...       ...                 ...              ...   
405619              0         0                 0.0                0   
405620              0         0                 0.0                0   
405621              3      1233               123.0                0   
405622              0         0                 0.0                0   
405623              0         0                 0.0                0   

        Dollar Cost Points Redeemed  
0                                 0  
1                                 0  
2                                 0  
3                                 0  
4                                 0  
...                             ...  
405619                            0  
405620                            0  
405621                            0  
405622                            0  
405623                            0  

[405624 rows x 10 columns]

In [23]:
df_activity.tail()

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
405619          999902  2018     12               0                        0   
405620          999911  2018     12               0                        0   
405621          999940  2018     12               3                        0   
405622          999982  2018     12               0                        0   
405623          999986  2018     12               0                        0   

        Total Flights  Distance  Points Accumulated  Points Redeemed  \
405619              0         0                 0.0                0   
405620              0         0                 0.0                0   
405621              3      1233               123.0                0   
405622              0         0                 0.0                0   
405623              0         0                 0.0                0   

        Dollar Cost Points Redeemed  
405619                            0  
405620                            0  
405621                            0  
405622                            0  
405623                            0

In [ ]:
df_activity.sample(5) # al ejecutar varias veces la celda, que las columnas de Points Redeemed y Dollar Cost Points Redeemed sus valores son 0 en casi todas las filas

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
237174          706962  2017      4               6                        6   
27937           160779  2018      5               0                        0   
190646          486860  2018     10               3                        2   
139119          309727  2017      9               6                        0   
29820           785223  2017      2               4                        0   

        Total Flights  Distance  Points Accumulated  Points Redeemed  \
237174             12      2496               249.0              563   
27937               0         0                 0.0                0   
190646              5      2610               261.0              588   
139119              6       696                69.0                0   
29820               4      1440               144.0                0   

        Dollar Cost Points Redeemed  
237174                           46  
27937                             0  
190646                           48  
139119                            0  
29820                             0

In [ ]:
df_activity.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 405624 entries, 0 to 405623
Data columns (total 10 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               405624 non-null  int64  
 1   Year                         405624 non-null  int64  
 2   Month                        405624 non-null  int64  
 3   Flights Booked               405624 non-null  int64  
 4   Flights with Companions      405624 non-null  int64  
 5   Total Flights                405624 non-null  int64  
 6   Distance                     405624 non-null  int64  
 7   Points Accumulated           405624 non-null  float64
 8   Points Redeemed              405624 non-null  int64  
 9   Dollar Cost Points Redeemed  405624 non-null  int64  
dtypes: float64(1), int64(9)
memory usage: 30.9 MB


In [ ]:
df_history.isna().sum() 
#ojo a las columnas CancellationYear y Month, que tienen ambas 14670 nulos, frente al total de 16737! (en esas columnas solo hay 2067 valores no nulos)
# Salary también tiene un recuento de nulos significativo. 

Loyalty Number            0
Country                   0
Province                  0
City                      0
Postal Code               0
Gender                    0
Education                 0
Salary                 4238
Marital Status            0
Loyalty Card              0
CLV                       0
Enrollment Type           0
Enrollment Year           0
Enrollment Month          0
Cancellation Year     14670
Cancellation Month    14670
dtype: int64

In [ ]:
round(df_history.isna().sum()/df_history.shape[0]*100,2)


Loyalty Number         0.00
Country                0.00
Province               0.00
City                   0.00
Postal Code            0.00
Gender                 0.00
Education              0.00
Salary                25.32
Marital Status         0.00
Loyalty Card           0.00
CLV                    0.00
Enrollment Type        0.00
Enrollment Year        0.00
Enrollment Month       0.00
Cancellation Year     87.65
Cancellation Month    87.65
dtype: float64

In [40]:
df_activity.isna().sum()

Loyalty Number                 0
Year                           0
Month                          0
Flights Booked                 0
Flights with Companions        0
Total Flights                  0
Distance                       0
Points Accumulated             0
Points Redeemed                0
Dollar Cost Points Redeemed    0
dtype: int64

In [ ]:
round(df_activity.isna().sum()/df_activity.shape[0]*100,2)
# No hay valores nulos!

Loyalty Number                 0.0
Year                           0.0
Month                          0.0
Flights Booked                 0.0
Flights with Companions        0.0
Total Flights                  0.0
Distance                       0.0
Points Accumulated             0.0
Points Redeemed                0.0
Dollar Cost Points Redeemed    0.0
dtype: float64

In [42]:
df_activity.describe().T

count           mean            std       min  \
Loyalty Number               405624.0  550037.873084  258935.286969  100018.0   
Year                         405624.0    2017.500000       0.500001    2017.0   
Month                        405624.0       6.500000       3.452057       1.0   
Flights Booked               405624.0       4.115052       5.225518       0.0   
Flights with Companions      405624.0       1.031805       2.076869       0.0   
Total Flights                405624.0       5.146858       6.521227       0.0   
Distance                     405624.0    1208.880059    1433.155320       0.0   
Points Accumulated           405624.0     123.692721     146.599831       0.0   
Points Redeemed              405624.0      30.696872     125.486049       0.0   
Dollar Cost Points Redeemed  405624.0       2.484503      10.150038       0.0   

                                   25%       50%        75%       max  
Loyalty Number               326961.00  550834.0  772194.00  999986.0  
Year                           2017.00    2017.5    2018.00    2018.0  
Month                             3.75       6.5       9.25      12.0  
Flights Booked                    0.00       1.0       8.00      21.0  
Flights with Companions           0.00       0.0       1.00      11.0  
Total Flights                     0.00       1.0      10.00      32.0  
Distance                          0.00     488.0    2336.00    6293.0  
Points Accumulated                0.00      50.0     239.00     676.5  
Points Redeemed                   0.00       0.0       0.00     876.0  
Dollar Cost Points Redeemed       0.00       0.0       0.00      71.0

In [43]:
df_history.describe().T

count           mean            std        min  \
Loyalty Number      16737.0  549735.880445  258912.132453  100018.00   
Salary              12499.0   79245.609409   35008.297285  -58486.00   
CLV                 16737.0    7988.896536    6860.982280    1898.01   
Enrollment Year     16737.0    2015.253211       1.979111    2012.00   
Enrollment Month    16737.0       6.669116       3.398958       1.00   
Cancellation Year    2067.0    2016.503145       1.380743    2013.00   
Cancellation Month   2067.0       6.962748       3.455297       1.00   

                          25%        50%        75%        max  
Loyalty Number      326603.00  550434.00  772019.00  999986.00  
Salary               59246.50   73455.00   88517.50  407228.00  
CLV                   3980.84    5780.18    8940.58   83325.38  
Enrollment Year       2014.00    2015.00    2017.00    2018.00  
Enrollment Month         4.00       7.00      10.00      12.00  
Cancellation Year     2016.00    2017.00    2018.00    2018.00  
Cancellation Month       4.00       7.00      10.00      12.00

##### - CLV tiene un count de 16737, sin nulos. Lo destacable es que tiene un valor max de 83325.38 mucho más elevado que el 75% de los datos(8940.58). ¿Outliers?
##### - La columna salario habrá que limpiarla, porque devuelve datos raros: Su recuento es de 12499, tiene 4238 valores nulos. ¿Gente que no percibe salario?; el valor mínimo es NEGATIVO ¿error de registro?; el valor máximo es mucho más alto que el 75%, posibles outliers en columna salario.
##### - Observaciones no significativas para análisis actual, en documentación.

In [48]:
df_history.describe(include="O").T  # en este df incluimos los objetos, para que nos devuelva un recuento de lo que no está calculando en el describe anterior que solo computa las variables numéricas.
#¿Aerolínea canadiense maybe?

count unique       top   freq
Country          16737      1    Canada  16737
Province         16737     11   Ontario   5404
City             16737     29   Toronto   3351
Postal Code      16737     55   V6E 3D9    911
Gender           16737      2    Female   8410
Education        16737      5  Bachelor  10475
Marital Status   16737      3   Married   9735
Loyalty Card     16737      3      Star   7637
Enrollment Type  16737      2  Standard  15766

In [46]:
df_activity.duplicated().sum() # para ver si tenemos duplicados en el df
# Duplicado por DataFrame busca filas completas repetidas.

np.int64(1864)

In [ ]:
df_history.duplicated().sum() # para ver si tenemos duplicados en el df
# Este output significa que no hay FILAS COMPLETAS repetidas en el df (eso no significa que no pueda haber valores repetidos en cada columna concreta)

np.int64(0)

In [ ]:
df_history["Country"].duplicated().sum() # por ejemplo, ésto es lógico y no es un problema de duplicado, si no que es un patrón, dado que concluimos que el país de toda la clientela era Canadá.

np.int64(16736)

In [54]:
def duplicados_columnas(df):
    total_filas = len(df)
    resultados = {}

    for columna in df.columns:
        conteo_repeticiones = df[columna].duplicated().sum()
        conteo_unicos = total_filas - conteo_repeticiones
        porcentaje_rep = (conteo_repeticiones / total_filas) * 100

        resultados[columna] = {
            'V-únicos': conteo_unicos,
            'V-repetidos': conteo_repeticiones,
            'Reps.porcentaje': f'{porcentaje_rep:.2f}%'
        }
    return pd.DataFrame(resultados).T

resultados_conteos = duplicados_columnas(df_history)
print(resultados_conteos)

                   V-únicos V-repetidos Reps.porcentaje
Loyalty Number        16737           0           0.00%
Country                   1       16736          99.99%
Province                 11       16726          99.93%
City                     29       16708          99.83%
Postal Code              55       16682          99.67%
Gender                    2       16735          99.99%
Education                 5       16732          99.97%
Salary                 5891       10846          64.80%
Marital Status            3       16734          99.98%
Loyalty Card              3       16734          99.98%
CLV                    7984        8753          52.30%
Enrollment Type           2       16735          99.99%
Enrollment Year           7       16730          99.96%
Enrollment Month         12       16725          99.93%
Cancellation Year         7       16730          99.96%
Cancellation Month       13       16724          99.92%


In [55]:
def duplicados_columnas(df):
    total_filas = len(df)
    resultados = {}

    for columna in df.columns:
        conteo_repeticiones = df[columna].duplicated().sum()
        conteo_unicos = total_filas - conteo_repeticiones
        porcentaje_rep = (conteo_repeticiones / total_filas) * 100

        resultados[columna] = {
            'V-únicos': conteo_unicos,
            'V-repetidos': conteo_repeticiones,
            'Reps.porcentaje': f'{porcentaje_rep:.2f}%'
        }
    return pd.DataFrame(resultados).T

resultados_conteos = duplicados_columnas(df_activity)
print(resultados_conteos)

                            V-únicos V-repetidos Reps.porcentaje
Loyalty Number                 16737      388887          95.87%
Year                               2      405622         100.00%
Month                             12      405612         100.00%
Flights Booked                    22      405602          99.99%
Flights with Companions           12      405612         100.00%
Total Flights                     33      405591          99.99%
Distance                        4746      400878          98.83%
Points Accumulated              1549      404075          99.62%
Points Redeemed                  587      405037          99.86%
Dollar Cost Points Redeemed       49      405575          99.99%


#### - Revisión resultados en Documentación.

## Unión datasets:

### Primero, limpieza de los datasets pre-mergeo, por eficiencia:

In [ ]:
# 1. Gestión registro anómalo, salario negativo. Convertir a nulo, para agruparlo con el resto de nulos de la columna y gestionarlos a posteriori juntos.
df_history.loc[df_history['Salary'] <=0, 'Salary'] = np.nan


In [58]:
# verificación de la conversión a nulo
print(f"El valor mínimo de Salary ahora es: {df_history['Salary'].min():,.2f}")

El valor mínimo de Salary ahora es: 15,609.00


In [ ]:
print(f"Ahora el conteo de valores nulos en Salary es: {df_history['Salary'].isnull().sum()}")

# dado que antes este recuento sumaba 4238, la conversión ha funcionado! 

Ahora el conteo de valores nulos en Salary es: 4258


In [ ]:
# 2. Creación de una columna binaria, por asignación directa, que permita eliminar las columnas de Year y Month que tienen unos tantos por cientos elevadísimos de nulos.
# Sustituir esas dos columnas que no aportan información de calidad (es información escasa e irrelevante) por una sola columna que conteste realmente a la pregunta de si la suscripción del cliente está activa o no.
# Ésto redunda en la eficiencia y rapidez del proceso de unión.

df_history['Status'] = df_history['Cancellation Year'].notna().astype(int)  # uso la columna cancellation year, en vez de la de month porque es más fiable (tiene 12 valores únicos, no como la de cancellation month que tiene 13: 12 meses+ NaN)


In [61]:
df_history.columns

Index(['Loyalty Number', 'Country', 'Province', 'City', 'Postal Code',
       'Gender', 'Education', 'Salary', 'Marital Status', 'Loyalty Card',
       'CLV', 'Enrollment Type', 'Enrollment Year', 'Enrollment Month',
       'Cancellation Year', 'Cancellation Month', 'Status'],
      dtype='object')

In [ ]:
df_history['Status'].value_counts() # Comprobación que columna nueva tiene valores 0 = No cancelado/ 1= Cancelado)

Status
0    14670
1     2067
Name: count, dtype: int64

In [ ]:
# Eliminación columnas innecesarias (estas 2 columnas pasan a ser ruido visual, porque ya tenemos el estado de la suscripción en 'Status')
df_history.drop(columns=['Cancellation Year', 'Cancellation Month'], inplace=True)

In [64]:
df_history.columns

Index(['Loyalty Number', 'Country', 'Province', 'City', 'Postal Code',
       'Gender', 'Education', 'Salary', 'Marital Status', 'Loyalty Card',
       'CLV', 'Enrollment Type', 'Enrollment Year', 'Enrollment Month',
       'Status'],
      dtype='object')

In [65]:
# Eliminación columnas innecesarias (si quisiéramos sacar el país de la clientela, que es por defecto, Canadá, podríamos usar las columnas de provincia, ciudad y código postal)
df_history.drop(columns= ['Country'], inplace=True)

In [66]:
df_history.columns

Index(['Loyalty Number', 'Province', 'City', 'Postal Code', 'Gender',
       'Education', 'Salary', 'Marital Status', 'Loyalty Card', 'CLV',
       'Enrollment Type', 'Enrollment Year', 'Enrollment Month', 'Status'],
      dtype='object')

In [70]:
df_history.shape

(16737, 14)

### Unión datasets:

In [67]:
df_fusionado = pd.merge(  # la función pd.merge() es para unir df's horizontalmente usando una columna común como eje de unión
    df_history,
    df_activity,
    on='Loyalty Number',
    how='left'
)

In [72]:
df_fusionado.columns

Index(['Loyalty Number', 'Province', 'City', 'Postal Code', 'Gender',
       'Education', 'Salary', 'Marital Status', 'Loyalty Card', 'CLV',
       'Enrollment Type', 'Enrollment Year', 'Enrollment Month', 'Status',
       'Year', 'Month', 'Flights Booked', 'Flights with Companions',
       'Total Flights', 'Distance', 'Points Accumulated', 'Points Redeemed',
       'Dollar Cost Points Redeemed'],
      dtype='object')

In [69]:
df_fusionado.shape

(405624, 23)

## Análisis post merge:

In [ ]:
# Repetir comprobación nulos, post mergeo:
df_fusionado.isnull().sum()

Loyalty Number                      0
Province                            0
City                                0
Postal Code                         0
Gender                              0
Education                           0
Salary                         103152
Marital Status                      0
Loyalty Card                        0
CLV                                 0
Enrollment Type                     0
Enrollment Year                     0
Enrollment Month                    0
Status                              0
Year                                0
Month                               0
Flights Booked                      0
Flights with Companions             0
Total Flights                       0
Distance                            0
Points Accumulated                  0
Points Redeemed                     0
Dollar Cost Points Redeemed         0
dtype: int64

#### El df_activity no presentaba nulos, y los del df_history estaban en las columnas eliminadas 'Cancellation Year/Month' y en 'Salary', que son lo que persisten y han aumentado (porque se han multiplicado por cada registro de cliente único)

In [74]:
df_fusionado

Loyalty Number  Province      City Postal Code  Gender Education  \
0               480934   Ontario   Toronto     M2Z 4K1  Female  Bachelor   
1               480934   Ontario   Toronto     M2Z 4K1  Female  Bachelor   
2               480934   Ontario   Toronto     M2Z 4K1  Female  Bachelor   
3               480934   Ontario   Toronto     M2Z 4K1  Female  Bachelor   
4               480934   Ontario   Toronto     M2Z 4K1  Female  Bachelor   
...                ...       ...       ...         ...     ...       ...   
405619          652627  Manitoba  Winnipeg     R2C 0M5  Female  Bachelor   
405620          652627  Manitoba  Winnipeg     R2C 0M5  Female  Bachelor   
405621          652627  Manitoba  Winnipeg     R2C 0M5  Female  Bachelor   
405622          652627  Manitoba  Winnipeg     R2C 0M5  Female  Bachelor   
405623          652627  Manitoba  Winnipeg     R2C 0M5  Female  Bachelor   

         Salary Marital Status Loyalty Card       CLV Enrollment Type  \
0       83236.0        Married         Star   3839.14        Standard   
1       83236.0        Married         Star   3839.14        Standard   
2       83236.0        Married         Star   3839.14        Standard   
3       83236.0        Married         Star   3839.14        Standard   
4       83236.0        Married         Star   3839.14        Standard   
...         ...            ...          ...       ...             ...   
405619  75049.0        Married         Star  83325.38        Standard   
405620  75049.0        Married         Star  83325.38        Standard   
405621  75049.0        Married         Star  83325.38        Standard   
405622  75049.0        Married         Star  83325.38        Standard   
405623  75049.0        Married         Star  83325.38        Standard   

        Enrollment Year  Enrollment Month  Status  Year  Month  \
0                  2016                 2       0  2017      1   
1                  2016                 2       0  2017      2   
2                  2016                 2       0  2017      3   
3                  2016                 2       0  2017      4   
4                  2016                 2       0  2018     10   
...                 ...               ...     ...   ...    ...   
405619             2015                12       1  2018      8   
405620             2015                12       1  2018      9   
405621             2015                12       1  2018     10   
405622             2015                12       1  2018     11   
405623             2015                12       1  2018     12   

        Flights Booked  Flights with Companions  Total Flights  Distance  \
0                    0                        0              0         0   
1                    3                        0              3      2823   
2                    0                        0              0         0   
3                    0                        0              0         0   
4                    6                        2              8      3352   
...                ...                      ...            ...       ...   
405619               0                        0              0         0   
405620               0                        0              0         0   
405621               0                        0              0         0   
405622               0                        0              0         0   
405623               0                        0              0         0   

        Points Accumulated  Points Redeemed  Dollar Cost Points Redeemed  
0                      0.0                0                            0  
1                    282.0                0                            0  
2                      0.0                0                            0  
3                      0.0                0                            0  
4                    335.0              465                           38  
...                    ...              ...                   

In [77]:
df_fusionado.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 405624 entries, 0 to 405623
Data columns (total 23 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               405624 non-null  int64  
 1   Province                     405624 non-null  object 
 2   City                         405624 non-null  object 
 3   Postal Code                  405624 non-null  object 
 4   Gender                       405624 non-null  object 
 5   Education                    405624 non-null  object 
 6   Salary                       302472 non-null  float64
 7   Marital Status               405624 non-null  object 
 8   Loyalty Card                 405624 non-null  object 
 9   CLV                          405624 non-null  float64
 10  Enrollment Type              405624 non-null  object 
 11  Enrollment Year              405624 non-null  int64  
 12  Enrollment Month             405624 non-null  int64  
 13 

In [ ]:
# 1. Elimina o trata los valores nulos, si los hay, en las columnas clave para asegurar que los datos estén completos:
# La única columna con nulos es'Salary'. Como argumenté en la documentación, vamos a imputarlos con la mediana (que es más robusta a los nulos, que la media), para que no altere las analíticas posteriores.

median_salary = df_fusionado['Salary'].median()
print(median_salary)

# no parece haberle afectado la eliminación de los outliers negativos, porque la media que resulta está muy cerca el percentil 50% que nos devolvían las estadísiticas descrptivas del describe()

73523.0


In [ ]:
# imputación de nulos con mediana
df_fusionado['Salary'] = df_fusionado['Salary'].fillna(median_salary)

In [ ]:
# 2. Verifica la consistencia y corrección de los datos para asegurarte de que los datos se presenten de forma coherente.

df_fusionado['Salary'].isnull().sum()

np.int64(0)

In [ ]:
# 3. Realiza cualquier ajuste o conversión necesaria en las columnas (por ejemplo, cambiar tipos de datos) para garantizar la adecuación de los datos para el análisis estadístico.
df_fusionado['Points Accumulated'].unique()

# contraintuitivamente, los puntos a acumular pueden ser (y son) con decimales.

array([  0.  , 282.  , 335.  , ...,  10.8 , 477.36, 661.5 ], shape=(1549,))

In [85]:
# Revisión de las 8 columnas categóricas, para detectar inconsistencias de texto:

columnas_cat = ['Province', 'City', 'Postal Code', 'Gender', 'Education', 'Marital Status', 'Loyalty Card', 'Enrollment Type']

for col in columnas_cat:
    print(f"\nColumna: {col}")
    print(df_fusionado[col].value_counts())


Columna: Province
Province
Ontario                 130896
British Columbia        106848
Quebec                   79968
Alberta                  23400
Manitoba                 15984
New Brunswick            15432
Nova Scotia              12576
Saskatchewan              9936
Newfoundland              6312
Yukon                     2688
Prince Edward Island      1584
Name: count, dtype: int64

Columna: City
City
Toronto           81096
Vancouver         62520
Montreal          49920
Winnipeg          15984
Whistler          14040
Halifax           12576
Ottawa            12312
Edmonton          11808
Quebec City       11760
Trenton           11760
Dawson Creek      10776
Fredericton       10344
Regina             9936
Kingston           9720
Tremblant          9600
Victoria           9504
Hull               8688
West Vancouver     7872
St. John's         6312
Thunder Bay        6264
Sudbury            5520
Moncton            5088
Calgary            4584
Banff              4296
London   

#### Dado los enunciados de las visualizaciones requeridas, hay que transformar algunas variables categóricas en números, para poder usarlas.

In [86]:
# 'Gender' es una columna binaria female/male, hay que codificarla 0/1:
df_fusionado['Gender'] = df_fusionado['Gender'].map({'Female': 1, 'Male': 0})


## Visualizaciones: 

In [ ]:
def archivar_grafica(nombre):
    ruta = f"images/{nombre}.png"
    plt.savefig(ruta, dpi=300, bbox_inches="tight")
    print(f" Gráfica archivada en: {ruta}")

In [ ]:
# NO EJECUTAR AÚN! 
# FASE DE VISUALIZACIONES (pegar esto cada vez que haga una gráfica):

# código para pegar al crear gráfica y que se ejecuten conjuntamente, para archivar directamente la imagen.
plt.figure(figsize=(10,6))
sns.barplot(data=df, x="Month", y="Flights Booked")
plt.title("Vuelos por mes")

plt.plot(x, y)
archivar_grafica("grafica_ejemplo")
plt.show()